hathor/indexes/addresses_rocksdb_index.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Iterable, List, Optional, Set, Tuple
+
+# TODO How to gracefully fail when rocksdb is not installed?
+import rocksdb
+from structlog import get_logger
+
+from hathor.indexes.addresses_index import AddressesIndex
+from hathor.pubsub import HathorEvents
+from hathor.transaction import BaseTransaction
+from hathor.transaction.scripts import parse_address_script
+
+if TYPE_CHECKING:  # pragma: no cover
+    from hathor.pubsub import EventArguments, PubSubManager
+    from hathor.transaction import TxOutput
+
+logger = get_logger()
+
+_CF_NAME_ADDRESS_INDEX = b'address-index'
+
+
+class AddressesRocksDBIndex(AddressesIndex):
+    """ Index of inputs/outputs by address.
+
+    This index uses rocksdb and the following key format:
+
+        key = [address][tx.timestamp][tx.hash]
+              |--34b--||--4 bytes---||--32b--|
+
+    It works nicely because rocksdb uses a tree sorted by key under the hoods.
+
+    The timestamp must be serialized in big-endian, so ts1 > ts2 implies that bytes(ts1) > bytes(ts2),
+    hence the transactions are sorted by timestamp.
+    """
+    def __init__(self, db: rocksdb.DB, *, cf_name: Optional[bytes] = None,
+                 pubsub: Optional['PubSubManager'] = None) -> None:
+        self._db = db
+        self._cf_name = cf_name or _CF_NAME_ADDRESS_INDEX
+        self._cf = self._db.get_column_family(self._cf_name)
```

Comment:
Also, I think this might be an OK place to reset any existing data. If we don't reset it won't work with an existing database and the current set of changes. To me resetting would be the only blocker to merging this PR.

Reply:
You mean to reset the memory index data?

Reply:
I've implemented data-resetting (by dropping and creating a new column-family) for the rocksdb-address-index

Reply:
If we continue with this, any restart would also need to reload all index again, right? We are not using the already saved index in storage to speed up sync. Or did I miss something?

Reply:
Yes, that's exactly it. This is done on `self._reset_cf()`, more precisely on `self._db.drop_column_family(_cf)`. Marcelo and I researched the performance implications and the conclusion is that the cost is basically O(1) and the data will be freed seamlessly.